File: lib/features/sessions/domain/streak_service.dart
import 'package:ten_k_hours/features/sessions/domain/session.dart';
import 'package:ten_k_hours/features/sessions/domain/streaks.dart';

class StreakService {
  const StreakService();

  // Compute per-pursuit streaks from counted sessions.
  //
  // - currentDays: consecutive day-run anchored at *today* if today has a
  //   session, otherwise anchored at *yesterday* (giving the user the full
  //   day to keep the streak alive). 0 if neither today nor yesterday has
  //   a session.
  // - longestDays: longest consecutive day-run anywhere in history.
  //
  // Session start times are domain-layer UTC; we project them to local
  // calendar dates using `nowLocal`'s timezone offset.
  Streaks compute({
    required Iterable<Session> countedSessions,
    required DateTime nowLocal,
  }) {
    final days = countedSessions
        .map((s) => _localDateOf(s.startedAt, nowLocal.timeZoneOffset))
        .toSet();
    if (days.isEmpty) return Streaks.empty;

    final today = _dateOnly(nowLocal);
    final yesterday = today.subtract(const Duration(days: 1));

    var current = 0;
    if (days.contains(today)) {
      current = _countBackFrom(today, days);
    } else if (days.contains(yesterday)) {
      current = _countBackFrom(yesterday, days);
    }

    return Streaks(currentDays: current, longestDays: _longestRun(days));
  }

  DateTime _localDateOf(DateTime instant, Duration offset) {
    final shifted = instant.toUtc().add(offset);
    return DateTime.utc(shifted.year, shifted.month, shifted.day);
  }

  DateTime _dateOnly(DateTime local) =>
      DateTime.utc(local.year, local.month, local.day);

  int _countBackFrom(DateTime anchor, Set<DateTime> days) {
    var count = 0;
    var d = anchor;
    while (days.contains(d)) {
      count++;
      d = d.subtract(const Duration(days: 1));
    }
    return count;
  }

  int _longestRun(Set<DateTime> days) {
    final sorted = days.toList()..sort();
    var best = 1;
    var run = 1;
    for (var i = 1; i < sorted.length; i++) {
      final delta = sorted[i].difference(sorted[i - 1]).inDays;
      if (delta == 1) {
        run++;
        if (run > best) best = run;
      } else {
        run = 1;
      }
    }
    return best;
  }
}

File: Sources/Models/StreakService.swift
import Foundation

// Pure streak computation from counted sessions. Per-pursuit.
// Ported from sessions/domain/streak_service.dart.
//
// - currentDays: consecutive day-run anchored at *today* if today has a session,
//   otherwise at *yesterday* (giving the user the full day to keep it alive).
//   0 if neither today nor yesterday has a session.
// - longestDays: longest consecutive day-run anywhere in history.
//
// Sessions are bucketed by local calendar date using `calendar` (its timeZone).
struct StreakService {
  func compute(
    countedSessions: [Session],
    now: Date,
    calendar: Calendar = .current
  ) -> Streaks {
    let days = Set(countedSessions.map { calendar.startOfDay(for: $0.startedAt) })
    guard !days.isEmpty else { return .empty }

    let today = calendar.startOfDay(for: now)
    let yesterday = calendar.date(byAdding: .day, value: -1, to: today)!

    var current = 0
    if days.contains(today) {
      current = countBack(from: today, in: days, calendar: calendar)
    } else if days.contains(yesterday) {
      current = countBack(from: yesterday, in: days, calendar: calendar)
    }

    return Streaks(currentDays: current, longestDays: longestRun(days, calendar: calendar))
  }

  private func countBack(from anchor: Date, in days: Set<Date>, calendar: Calendar) -> Int {
    var count = 0
    var d = anchor
    while days.contains(d) {
      count += 1
      d = calendar.date(byAdding: .day, value: -1, to: d)!
    }
    return count
  }

  private func longestRun(_ days: Set<Date>, calendar: Calendar) -> Int {
    let sorted = days.sorted()
    var best = 1
    var run = 1
    for i in 1..<sorted.count {
      let delta = calendar.dateComponents([.day], from: sorted[i - 1], to: sorted[i]).day ?? 0
      if delta == 1 {
        run += 1
        best = max(best, run)
      } else {
        run = 1
      }
    }
    return best
  }
}
